### PR TITLE
Rewrite User Guide workflow and tool (feature) pages.

### PR DESCRIPTION
**Pull request by dwelsch-esi**
Revised to include strategy guidelines and extended explanations.

### CODE REVIEW

File: Docs/userguide/quantization_workflow.rst [hunk 1]
```diff
@@ -1,158 +1,77 @@
...
-
-W16A16 verification
--------------------
+Step 3. Reducing precision
+--------------------------
 
-Before using quantized integer format, it's important to ensure that the FP32 model
-and the quantized model (QuantSim object) perform similarly during the forward pass, especially
-when custom quantizers are included in the model.
+To gain on-target performance, you have to reduce precision. We suggest starting with with weights at INT8 precision and activations at INT16 precision (W8A16). Again, the aim is to reduce precision as little as possible to achieve acceptable performance. There's nothing keeping you from trying a more radical reduction in precision if you want, however. Integer precisions supported in AIMET are listed in :ref:`supported_on_target_precisions`.
```
Comment (quic-kyunggeu) [inline Docs/userguide/quantization_workflow.rst:63]: I think the term "performance" is unfortunately badly abused in computer science. It can mean accuracy or speed/energy/etc.

I was assuming "performance" meant accuracy, and I had to think about this sentence for quite a while until I realized "performance" meant speed/energy/etc

> Again, the aim is to reduce precision as little as possible to achieve acceptable performance

Because if you interpret "performance" as accuracy, you would want to reduce precision as **much** as possible.

I don't have a concrete suggestion though. Maybe "runtime performance" will make it sound less ambiguous?

File: Docs/userguide/quantization_workflow.rst [hunk 2]
```diff
@@ -1,158 +1,77 @@
...
+Do the following:
 
-Given that the off-target quantized accuracy using QuantSim is acceptable, following
-on-target metrics should be considered depending on your application.
+1. Ensure that the FP32 model adheres to model-specific guidelines. For instance, in PyTorch QuantSim can only quantize math operations performed by :class:`torch.nn.Module` objects, while :class:`torch.nn.functional` calls will be incorrectly ignored. See framework-specific pages to learn more about such model guidelines.
+2. Once the model conforms to guidelines, create a quantization simulation (QuantSim) version of your model with the bit-width set to 16 bits for both weights and activations (W16A16). See :ref:`<quantsim-workflow>`.
+3. Ensure that the original FP32 model and the quantized model (QuantSim object) perform similarly during the forward pass, especially when custom quantizers are included in the model.
+4. Compute the off-target quantized accuracy metric for the quantized model and verify that it agrees (approximately) with the FP32 model. If it does not, there might be a problem with the quantization nodes. You can help improve AIMET by reporting an issue to |aimet|_.
```
Comment (quic-kyunggeu) [inline Docs/userguide/quantization_workflow.rst]: > Ensure that the FP32 model adheres to model-specific guidelines

You intended "framework-specific", right?


> 3. Ensure that the original FP32 model and the quantized model (QuantSim object) perform similarly during the forward pass, especially when custom quantizers are included in the model.

I'm not fully confident about the "custom quantizers" part.

@quic-akhobare  @quic-hitameht
Is this in consideration of something that we plan to support in the future? We don't support custom quantizers today, do we?

**Review by quic-kyunggeu — DISMISSED** [review]

File: Docs/userguide/quantization_workflow.rst (hunk 2 above)
Comment (dwelsch-esi) [reply Docs/userguide/quantization_workflow.rst]: @quic-kyunggeu
"framework-specific", yes.

"Custom quantizers" -- that was in the original 2.0 rewrite that I was given. Happy to remove it if it's not true.

File: Docs/userguide/quantization_workflow.rst [hunk 3]
```diff
@@ -1,158 +1,77 @@
...
+Add quantization/dequantization (QDQ) nodes to the model using QuantSim. Set weights and activations to 16-bit integer (W16A16). This simulated quantization in the modified floating-point model is called *off-target* quantization. 
 
-To decide which precision to run inference on target runtimes, you can follow the top-down
-approach where you begin with the highest precision (For example FP16) and transition to
-lower precision if necessary, which may require additional engineering effort.
+Do the following:
 
-Given that the off-target quantized accuracy using QuantSim is acceptable, following
-on-target metrics should be considered depending on your application.
+1. Ensure that the FP32 model adheres to model-specific guidelines. For instance, in PyTorch QuantSim can only quantize math operations performed by :class:`torch.nn.Module` objects, while :class:`torch.nn.functional` calls will be incorrectly ignored. See framework-specific pages to learn more about such model guidelines.
```
Comment (dwelsch-esi) [inline Docs/userguide/quantization_workflow.rst]:
```suggestion
1. Ensure that the FP32 model adheres to framework-specific guidelines. For instance, in PyTorch QuantSim can only quantize math operations performed by :class:`torch.nn.Module` objects, while :class:`torch.nn.functional` calls will be incorrectly ignored. See framework-specific pages to learn more about such model guidelines.
```

File: Docs/userguide/quantization_workflow.rst (hunk 1 above)
Comment (dwelsch-esi) [reply Docs/userguide/quantization_workflow.rst:63]: @quic-kyunggeu 
Good point. In my experience, "performance" usually refers to some aspect of speed: ops per second, bandwidth, throughput, etc. But not always.

The original workflow writeup specified
```
-    Latency reduction and/or
-   Memory size reduction
```
as the on-target performance metrics. I'll add that back in and make it clear.

This is also something that could go in the glossary if it's decided to restrict the definition of "performance" for AIMET.

File: Docs/userguide/quantization_workflow.rst (hunk 2 above)
Comment (quic-kyunggeu) [reply Docs/userguide/quantization_workflow.rst]: "Custom quantizers" -- Then I'll have to talk with the original author of this paragraph. It looks good to me otherwise

**Review by quic-kyunggeu — APPROVED** [review]

File: Docs/userguide/quantization_workflow.rst (hunk 2 above)
Comment (quic-hitameht) [reply Docs/userguide/quantization_workflow.rst]: AIMET added QDQ operations are referred to as `custom quantizers`.

**Review by quic-hitameht — DISMISSED** [review]

File: Docs/userguide/quantization_workflow.rst (hunk 2 above)
Comment (quic-kyunggeu) [reply Docs/userguide/quantization_workflow.rst]: @quic-hitameht I see. IMO that is a bit misleading though... "Custom" sounds to me as if the user implemented it by themselves.
Personally, I couldn't realize that "custom quantizers" are AIMET quantizers before I searched for all other usages of "custom quantizers" in other pages of the doc.
When we mention "custom quantizers" in other pages, it wasn't so confusing because they kindly recap its definition. But in this page, it's quite confusing to me.

Maybe can we just remove "especially when ..." part? It seems like a needless nuance to me. Even if I interpreted the meaning of "custom quantizer" correctly, it still sounds as if the QuantSim model may not contain any quantizer in some major use cases.

File: Docs/userguide/quantization_workflow.rst [hunk 4]
```diff
@@ -1,158 +1,77 @@
...
-To decide which precision to run inference on target runtimes, you can follow the top-down
-approach where you begin with the highest precision (For example FP16) and transition to
-lower precision if necessary, which may require additional engineering effort.
+Do the following:
 
-Given that the off-target quantized accuracy using QuantSim is acceptable, following
-on-target metrics should be considered depending on your application.
+1. Ensure that the FP32 model adheres to framework-specific guidelines. For instance, in PyTorch QuantSim can only quantize math operations performed by :class:`torch.nn.Module` objects, while :class:`torch.nn.functional` calls will be incorrectly ignored. See framework-specific pages to learn more about such model guidelines.
+2. Once the model conforms to guidelines, create a quantization simulation (QuantSim) version of your model with the bit-width set to 16 bits for both weights and activations (W16A16). See :ref:`<quantsim-workflow>`.
+3. Ensure that the original FP32 model and the quantized model (QuantSim object) perform similarly during the forward pass, especially when custom quantizers are included in the model.
```
Comment (quic-kyunggeu) [inline Docs/userguide/quantization_workflow.rst]:
```suggestion
3. Ensure that the original FP32 model and the quantized model (QuantSim object) perform similarly during the forward pass.
```

@dwelsch-esi  @quic-hitameht suggestion from [this comment](https://github.com/quic/aimet/pull/3796#discussion_r1951337166)

**Review by quic-kyunggeu — DISMISSED** [review]

File: Docs/userguide/quantization_workflow.rst (hunk 4 above)
Comment (quic-hitameht) [reply Docs/userguide/quantization_workflow.rst]: Yes, I am fine with removing the "Especially part". @dwelsch-esi can you please address this?

Comment (dwelsch-esi) [reply Docs/userguide/quantization_workflow.rst]: Removed in the revision I'm working on.

File: Docs/userguide/quantization_workflow.rst (hunk 1 above)
Comment (dwelsch-esi) [reply Docs/userguide/quantization_workflow.rst:63]: @quic-kyunggeu @quic-hitameht How about if we say something like this in the introduction?


  General guidelines
  ==================
  
  Preparing a model for deployment on a target device involves a tradeoff between model *accuracy* and on-device *performance*. 
  
  Accuracy is how "correct" the model response is, typically a fraction of responses correct as measured by an evaluation function.
  
  Performance, in an AIMET context, is the ability of a given device to run a model without overtaxing the device compute and memory resources. To evaluate performance improvements, we recommend measuring latency reduction and memory size reduction.

**Review by quic-kyunggeu — APPROVED** [review]
